app/models/chapter.rb

```diff
@@ -45,4 +45,36 @@ def display_with_mlid
   def full_mlid
     "#{organization.mlid}-#{mlid}"
   end
+
+  def delete_chapter_and_dependents
+    transaction do
+      self.deleted_at = Time.zone.now
+
+      # rubocop:disable Rails/SkipsModelValidations
+      groups_to_update = Group.includes(:chapter).where(chapter_id: id, deleted_at: nil)
+      group_ids = groups_to_update.pluck(:id)
+
+      groups_to_update.update_all(deleted_at:)
+      Student.includes(:group).where(groups: group_ids, deleted_at: nil).update_all(deleted_at:)
+      Lesson.includes(:group).where(groups: group_ids, deleted_at: nil).update_all(deleted_at:)
+      Grade.includes(:lesson).where(lessons: { group_id: group_ids, deleted_at: }, deleted_at: nil).update_all(deleted_at:)
+      # rubocop:enable Rails/SkipsModelValidations
+    end
+  end
+
+  def restore_chapter_and_dependents
+    transaction do
+      # rubocop:disable Rails/SkipsModelValidations
+      groups_to_update = Group.includes(:chapter).where(chapter_id: id, deleted_at:)
+      group_ids = groups_to_update.pluck(:id)
```

Comment:
This is good since pluck does everything in the database. Much better than fetching all and then extracting ids in ruby. 👍